### ISSUE TITLE
Difference to the original implementation of the KL loss

### ISSUE THREAD

**Issue by jan1854** [ku2482/slac.pytorch#9]
Hi,

I think there is a slight discrepancy to the original implementation of the calculation of the KL loss. Here, each prior state used in the KL loss is calculated from the previous *prior* state.
https://github.com/ku2482/slac.pytorch/blob/d68f42582180d38735496a24cddbf0523726973d/slac/network/latent.py#L263-L266

In the original implementation, it seems like each prior state is calculated from the previous *posterior* state (see https://github.com/alexlee-gk/slac/blob/14a366e6a098165318a3b90fc50ff7e353cd43b4/slac/agents/slac/model_distribution_network.py#L287-L288).
This is also more consistent with the loss defined in equation 6 of the SLAC paper where z<sub>t</sub> is drawn from the posterior distribution q.

Best regards,
Jan

**Comment by toshikwa** [comment]
Hi @JanS97 

Actually, I calculated samples from prior distributions here, which should be consistent with the paper. 
I haven't read the original impl well, but I think it corresponds to [this part](https://github.com/alexlee-gk/slac/blob/14a366e6a098165318a3b90fc50ff7e353cd43b4/slac/agents/slac/model_distribution_network.py#L285-L304). 

https://github.com/ku2482/slac.pytorch/blob/d68f42582180d38735496a24cddbf0523726973d/slac/network/latent.py#L191-L210

Does it make sense?

Thanks.

**Comment by jan1854** [comment]
Hi,

if you have a close look at [lines 287-288](https://github.com/alexlee-gk/slac/blob/14a366e6a098165318a3b90fc50ff7e353cd43b4/slac/agents/slac/model_distribution_network.py#L287-L288) of the code snippet that you linked, you will see that the posterior states `latent2_posterior_samples` are fed into `latent1_prior()`. In your implementation, you always feed the last prior state into `z1_prior()`.

Best regards,
Jan

**Comment by toshikwa** [comment]
Hi,  thank you for your response!

SLAC uses the same distribution for prior and posterior of z2. ([original impl](https://github.com/alexlee-gk/slac/blob/14a366e6a098165318a3b90fc50ff7e353cd43b4/slac/agents/slac/model_distribution_network.py#L236-L243))
https://github.com/ku2482/slac.pytorch/blob/d68f42582180d38735496a24cddbf0523726973d/slac/network/latent.py#L156-L167

I'm sorry that my implementation is confusing, however it seems correct.
Anyway, thanks ;)

**Comment by jan1854** [comment]
Hi,

I am aware that the z<sup>2</sup> distribution is shared between the prior and posterior. However, `latent2_prior_samples` is still different from the prior since the z<sup>1</sup> that is used as input for the z<sup>2</sup> distribution is sampled according to the posterior distribution.
I tried to visualize the difference below. Note that I removed the actions for simplicity.

![slac_pytorch](https://user-images.githubusercontent.com/33448112/130211220-bee81974-b3d1-44f1-afee-903b1b9c00e6.png)
![original_slac](https://user-images.githubusercontent.com/33448112/130211185-04ca416b-6915-41ad-b851-306aee76ef87.png)

I hope that this makes it a bit clearer.

**Comment by toshikwa** [comment]
@JanS97 
Sorry for responding so late.

I realized what you were talking about.
We need to calculate the loss in the 1 step, not in the sequence of actual distributions.

I modified the code in this [commit](https://github.com/ku2482/slac.pytorch/commit/779711249643094b05094127d00e8a8c75f37029) (I'm evaluating now).
Could you please check if this is what you talked about?

Anyway, I really appreciate your advise.

**Comment by jan1854** [comment]
Yes, this is what I meant. Thanks for fixing it.

**Comment by y-ito-1985** [comment]
Thank you very much for your efforts in porting SLAC to PyTorch.
Even as I explore the latest research, I still find the elegance of this algorithm stands out.

Also, in the discussion above,
I found it very insightful to realize that in a VAE-based latent dynamics model,
the prior can be computed in a batch after computing the posterior.

I have one question.

In this implementation, only the KLD for z1 ​is computed,
whereas in the original implementation,
the KLD for both z1​ and z2​ is computed and added to the ELBO.
How should this difference be interpreted?

I would appreciate your comments on this.

// https://github.com/toshikwa/slac.pytorch/blob/master/slac/network/latent.py

@torch.jit.script_method
def calculate_loss(self, state_, action_, reward_, done_):
    ...
    loss_kld = calculate_kl_divergence(z1_mean_post_, z1_std_post_, z1_mean_pri_, z1_std_pri_).mean(dim=0).sum()
    ...

// https://github.com/alexlee-gk/slac/blob/master/slac/agents/slac/model_distribution_network.py

def compute_loss(self, images, actions, step_types, rewards=None, discounts=None, latent_posterior_samples_and_dists=None):
...
if self.kl_analytic:
  latent1_kl_divergences = tfd.kl_divergence(latent1_posterior_dists, latent1_prior_dists)
...
latent1_kl_divergences = tf.reduce_sum(latent1_kl_divergences, axis=1)
...
  if self.kl_analytic:
    latent2_kl_divergences = tfd.kl_divergence(latent2_posterior_dists, latent2_prior_dists)
...
  latent2_kl_divergences = tf.reduce_sum(latent2_kl_divergences, axis=1)
...
elbo = likelihood_log_probs - latent1_kl_divergences - latent2_kl_divergences
